### PR TITLE
lb between waypoint addresses

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/xds/endpoint_builder.go [hunk 1]
```diff
@@ -211,9 +212,9 @@ type LocalityEndpoints struct {
 	llbEndpoints endpoint.LocalityLbEndpoints
 }
 
-func (e *LocalityEndpoints) append(ep *model.IstioEndpoint, le *endpoint.LbEndpoint) {
+func (e *LocalityEndpoints) append(ep *model.IstioEndpoint, le ...*endpoint.LbEndpoint) {
```
Comment (howardjohn) [inline pilot/pkg/xds/endpoint_builder.go:215]: We have

```
func (e *LocalityEndpoints) AssertInvarianceInTest() {
	if len(e.llbEndpoints.LbEndpoints) != len(e.istioEndpoints) {
		panic(" len(e.llbEndpoints.LbEndpoints) != len(e.tunnelMetadata)")
	}
}
```

so this seems to violate that

Comment (hzxuzhonghu) [reply pilot/pkg/xds/endpoint_builder.go:215]: we need to update this assert to exclude waypoint case.

BTW, not sure why the integrate test failed

**Comment by hzxuzhonghu** [comment]
cc @kyessenov

**Comment by kyessenov** [comment]
Sorry, can you explain how LB works between waypoints? How can we load balance between tunnels?

**Comment by hzxuzhonghu** [comment]
Previously, eds only contains one waypoint, now all the waypoints are there.

**Comment by kyessenov** [comment]
@hzxuzhonghu Yeah, my concern is that waypoints require HBONE sessions, but they also represent a single endpoint.
For example, let's say you have an L7 pool:

1. regular IP.
2. HBONE session to waypoint pool for IP.

We want 50% requests go to waypoint-fronted endpoints. I don't understand how we can do this without creating a separate indirection with internal listeners and L4 pool for waypoints. Is that what's done here?

**Comment by hzxuzhonghu** [comment]
This is setting tunnel to waypoint:15008, and the target address keep as `connect_originate`.  So with this, we will have 

multi endpoints with  `connect_originate` address and `tunnel: waypoint adress`. Do you mean this can not work?


I do not understand what `HBONE sessions` is  here

**Comment by kyessenov** [comment]
I stared at the code and still have no idea what the effects are. Can you just post the xDS diff? How do you test?

Before, an endpoint is something like:
```
address:
  envoy_internal: connect_originate
metadata:
  tunnel:
    address: WAYPOINT_IP
    destination: POD_IP behind the waypoint.
```

What is it now?

**Comment by kyessenov** [comment]
Yeah, this cannot work - if you duplicate the above for the same POD_IP, then you disturb the weights. If you flip WAYPOINT_IP then you disturb long living connections/requests.

Waypoints are sessions, since we can't transfer inner streams between waypoints without abruptly killing them.

**Comment by hzxuzhonghu** [comment]
Yes, There can be multi with different WAYPOINT_IP but same POD_IP
```
address:
  envoy_internal: connect_originate
metadata:
  tunnel:
    address: WAYPOINT_IP
    destination: POD_IP behind the waypoint.
```

**Comment by kyessenov** [comment]
@hzxuzhonghu Random is not good either because it has to be consistent over xDS updates. What we really need is a two-level load balancer - first one is the standard EDS, and then a secondary one chooses a tunnel to that destination. That means our final listener `connect_originate` should use an EDS cluster as well, not ORIGINAL_DST. In other words, the right way to solve this is to have a dedicated listener for every waypoint (effectively VIP listener) and its own cluster.

**Comment by howardjohn** [comment]
@kyessenov what is the behavioral difference there? They seem equivalent to me

**Comment by kyessenov** [comment]
@howardjohn If you use only first EDS, you're 1) breaking the weights, if it's 50% to waypoints and 50% direct, it shouldn't matter how many waypoints are there; 2) losing waypoint load context, the load on the destination and the load on the waypoint shouldn't be coupled.

**Comment by howardjohn** [comment]
On the weights - agree but this can be solved by adjusting the other weights like we do for E-W gateway. Good point on the load - the same actually also probably applies to E-W gateways.

**Comment by kyessenov** [comment]
I don't like the multiplicative effect as well, for N waypoints and M backends, we create N*M endpoints in the cluster.

**Comment by kyessenov** [comment]
@howardjohn @hzxuzhonghu @costinm  If we feel like we need >1 waypoints, then I'm strongly against doing more hacks with weights distribution, and we should do it the right way to start with.

I think if we treat waypoints as another (headless?) service, we can also do better drain and migration, and align better with stateful sessions for sidecars. There's also an opportunity to drop a client ztunnel requirement by giving waypoints a VIP and addressing them from non-istio workloads.

**Comment by howardjohn** [comment]
Is the "right way" another internal listener, or tweaking Envoy to allow multi-level load balancing selection? I am a bit concerned about the former.

We could use a Service and just always have 1 endpoint, relying on k8s load balancing to pick the waypoint for us as well - would be much simpler

**Comment by kyessenov** [comment]
@howardjohn Not another listener on the chain. I'm suggesting changing from ORIGINAL_DST to EDS cluster for origination - that would imply more of them horizontally but not vertically. Or just use VIP as you suggested since we're doing TCP load balancing between waypoints. The advantage of EDS is "headless", but if we have VIPs to spare that is fine too.

**Comment by howardjohn** [comment]
One thing I was thinking is having a Service allows integrating non-istio things easier, we can just put the Service in the Gateway.status.address. But we can do that with headless a well.

So I think what sounds best to me is replace with EDS cluster(s) and use a headless Service. The latter is probably unrelated

**Comment by kyessenov** [comment]
For alpha, I think having VIP service for Waypoint is the best option, and putting that as Waypoint IP.

For beta, we can support headless waypoints and add EDS for VIP/VIP-less. The benefits are marginal for the complexity it incurs.

**Comment by costinm** [comment]
My primary requirement is to support external, managed waypoints. That means having a VIP - which may not be a pod, and may be shared among multiple namespaces, clusters or tenants.

I also think a reasonable requirement is to not require waypoint for a namespace to not be present in all clusters, to keep isolation. 

Using stateful set is an interesting idea, but not sure we have enough understanding of all implications. All for exploring them - but maybe long term and not as the master plan until we try a prototype and have deeper understanding.

**Comment by hzxuzhonghu** [comment]
Using service vip as the waypoint address has some disadvantages

1. cross-cluster can not work
2. if we want fine-grained control over destinations,it is hard
3. it violate we never depend on kube-proxy to do lb

**Comment by kyessenov** [comment]
@hzxuzhonghu none of these seem particularly problematic for alpha to me. I never heard of (3) being a requirement for Istio.

**Comment by hzxuzhonghu** [comment]
Actually, some users expect not to deploy kube-proxy at all with istio installed

**Comment by kyessenov** [comment]
@hzxuzhonghu Sure, there are lots of situations when ambient is inappropriate right now. It's an experiment, not something we're ready to replace sidecars with. Doing kube-proxy is clearly a ztunnel requirement, so even if we use VIP in waypoints, ztunnel should be doing L4 balancing on the upstream, not envoy.

**Comment by linsun** [comment]
> Sorry for joining the party late @hzxuzhonghu - why cross-cluster can not work with service VIP as waypoint addr?

Are you referring to scenarios where C (cluster1) calls S (cluster2) where S doesn't exist in cluster1?

**Review by howardjohn — COMMENTED** [review]
semi-related https://github.com/istio/istio/pull/43889

**Comment by hzxuzhonghu** [comment]
>Are you referring to scenarios where C (cluster1) calls S (cluster2) where S doesn't exist in cluster1?

@linsun Yeah, this is one reason

**Pull request closed without merging**